parser/src/parser.rs

```diff
@@ -0,0 +1,523 @@
+    map(
+        pair(
+            one_of(&[
+                Token::Keyword(Keyword::Void),
+                Token::Keyword(Keyword::Int),
+                Token::Keyword(Keyword::Unsigned),
+                Token::Keyword(Keyword::Bool),
+            ]),
+            fold_many0(tok(Token::Star), || 0, |s, _| s + 1),
+        ),
+        |(token, stars)| {
+            let mut typ = match token {
+                Token::Keyword(Keyword::Void) => Type::Void,
+                Token::Keyword(Keyword::Int) => Type::Int32,
+                Token::Keyword(Keyword::Unsigned) => Type::Unsigned32,
+                Token::Keyword(Keyword::Bool) => Type::Bool,
+                _ => unreachable!(),
+            };
+            for _ in 0..stars {
+                typ = Type::Ref(Box::new(typ))
+            }
+            typ
+        },
+    )(i)
+}
+
+/// Parse a specified token.
+fn tok(token: Token) -> impl Fn(&[Token]) -> IResult<&[Token], Token> {
+    move |i| {
+        map_res(take1, |first_token| {
+            if token == first_token {
+                Ok(token.clone())
+            } else {
+                Err(nom::Err::Error(NomError::from_error_kind(
+                    i,
+                    ErrorKind::Tag,
+                )))
+            }
+        })(i)
+    }
+}
+
+/// Parse an ident.
+fn ident(i: &[Token]) -> IResult<&[Token], Token> {
+    map_res(take1, |token| match token {
+        Token::Ident(_) => Ok(token),
+        _ => Err(nom::Err::Error(NomError::from_error_kind(
+            i,
+            ErrorKind::Tag,
+        ))),
+    })(i)
+}
+
+/// Parse a literal.
+fn literal(i: &[Token]) -> IResult<&[Token], Token> {
+    map_res(
+        pair(maybe(one_of(&[Token::Plus, Token::Minus])), take1),
+        |(sign, token)| match token {
+            Token::Number(n) => match sign {
+                Some(Token::Plus) | None => Ok(Token::Number(n)),
+                // TODO: overflow.
+                Some(Token::Minus) => Ok(Token::Number(-n)),
+                _ => unreachable!(),
+            },
+            _ => Err(nom::Err::Error(NomError::from_error_kind(
+                i,
+                ErrorKind::Tag,
+            ))),
+        },
+    )(i)
+}
+
+/// Parse a function signature.
+fn signature(i: &[Token]) -> IResult<&[Token], (Type, String, Vec<(Type, Variable)>)> {
+    map(
+        tuple((
+            // Actual grammar is here.
+            typ,
+            ident,
+            tok(LPar),
+            separated_list0(tok(Comma), pair(typ, ident)),
+            tok(RPar),
+        )),
+        |(return_type, name, _, decl, _)| {
+            (
+                return_type,
+                name.into_ident().unwrap(),
+                decl.into_iter()
+                    .map(|(typ, name)| {
+                        (
+                            typ.clone(),
+                            Variable {
+                                name: name.into_ident().unwrap().clone(),
```

Comment:
Why `clone` if we have already moved `decl` by calling `into_iter()`?
```suggestion
                            typ,
                            Variable {
                                name: name.into_ident().unwrap(),
```

Reply:
Applied.